File: Compiler/src/parsers/Variable.ts
import { ParserPointer, Token } from '../utils/ParserPointer';
import { SyntaxError } from '../errors/SyntaxError';
import { Expression } from './Expression';

export class Variable {
	constructor(private pointer: ParserPointer, private expression: Expression) {}

	variable() {
		const { pointer } = this;

		if (!pointer.token || !pointer.take('VariableKeyword')) return null;

		let last = 'let';

		const name = pointer.take('Identifier');
		if (!name) new SyntaxError(this.pointer, 'let', 'Expected a variable name', 'parser');
		last += ` ${name?.value}`;

		const assign = pointer.take('Assignment');
		if (!assign) {
			pointer.take('Semicolon');

			return {
				type: 'VariableDeclaration',
				name,
				value: 'undefined',
			};
		}
		last += `=`;

		const value = this.expression.expression();
		if (!value) new SyntaxError(this.pointer, last, 'Expected a variable value', 'parser');

		if (this.pointer.token?.type === 'Comma')
			new SyntaxError(this.pointer, ',', 'Unexpected comma', 'parser');

		pointer.take('Semicolon');

		return {
			type: 'VariableDeclaration',
			name,
			value,
		};
	}
}

File: Compiler/src/parsers/statements/variable.rs
use crate::utils::AvoidingBlock;
use std::mem::ManuallyDrop;

use crate::parsers::expression;
use crate::utils::{findName, Expression, Pointer, StatementToken, Token};

pub fn variable(
    pointer: &mut ManuallyDrop<Pointer>,
    body: &mut AvoidingBlock,
    isConstant: bool,
) -> Option<StatementToken> {
    pointer.take("Keyword", true, true);

    if let Some(Token::Identifier(name, _)) = pointer.take("Identifier", true, true) {
        let exists = findName(&body.current, name.clone());
        if exists.is_some() {
            pointer.error(format!("Identifier '{}' already declared", name));
        }

        if let Some(Token::Punctuation(punc, _)) = pointer.take("Punctuation", true, true) {
            if punc == "=" {
                let expr = expression(pointer);

                if let Some(expr) = expr {
                    if isConstant {
                        return Some(StatementToken::ConstantDeclaration(name, expr));
                    } else {
                        return Some(StatementToken::VariableDeclaration(name, expr));
                    }
                }

                pointer.error("Expected expression".to_string());
            }
        }

        if isConstant {
            pointer.error("Expected '='".to_string());

            return None;
        }

        return Some(StatementToken::VariableDeclaration(
            name,
            Expression::Undefined,
        ));
    }

    pointer.error(format!(
        "Expected a {} name",
        if isConstant { "constant" } else { "variable" }
    ));

    None
}
